File: 54-spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        n=len(matrix)
        m=len(matrix[0])

        top=0
        bottom=n-1
        left=0
        right=m-1
        res=[]

        while top<=bottom and left<=right:
            #left to right
            for i in range(left,right+1):
                res.append(matrix[top][i])
            top+=1

            # top -> bottom
            for i in range(top,bottom+1):
                res.append(matrix[i][right])
            right-=1

            # right to left
            if top<=bottom:
                for i in range(right,left-1,-1):
                    res.append(matrix[bottom][i])
                bottom-=1

            # bottom to top
            if left<=right:
                for i in range(bottom,top-1,-1):
                    res.append(matrix[i][left])
                left+=1
        return res

File: 54-spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        ArrayList<Integer> res=new ArrayList<>();
        
        int m=matrix.length;
        int n=matrix[0].length;

        int left=0,right=n-1;
        int top=0,bottom=m-1;

        while(top<=bottom && left<=right){

            //left to right
            for(int i=left;i<=right;i++){
                res.add(matrix[top][i]);
            }
            top++;


            //top to bottom
            for(int i=top;i<=bottom;i++){
                res.add(matrix[i][right]);
            }
            right--;

            //right to left
            if(top<=bottom){
                for(int i=right;i>=left;i--){
                    res.add(matrix[bottom][i]);
                }
                bottom--;
            }

            //bottom to top
            if(left<=right){
                for(int i=bottom;i>=top;i--){
                    res.add(matrix[i][left]);
                }
                left++;
            }

        }


        return res;




    }
}